fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,375 @@
+        .with_context(|| format!("Decoding meta field value '{str_value}' failed"));
+
+    // In the past we encoded some string fields as "just a string" without quotes,
+    // this code ensures that old meta values still parse since config is hard to
+    // change
+    if res.is_err() && std::any::TypeId::of::<V>() == std::any::TypeId::of::<String>() {
+        let string_ret = Box::new(str_value.to_owned());
+        let ret: Box<V> = unsafe {
+            // We can transmute a String to V because we know that V==String
+            std::mem::transmute(string_ret)
+        };
+        Ok(*ret)
+    } else {
+        res
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+pub trait MetaSource: MaybeSend + MaybeSync {
+    /// This must be stable.
+    fn source_id(&self) -> &'static str;
+    /// This must be stable.
+    fn priority(&self) -> u64;
+    /// Wait for next change in this source.
+    async fn wait_for_update(&self);
+    async fn fetch(
+        &self,
+        client: &Client,
+    ) -> anyhow::Result<BTreeMap<MetaFieldName, MetaFieldValue>>;
+}
+
+/// Service for managing the caching of meta fields.
+#[derive(Default)]
+pub struct MetaService {
+    sources: Vec<Box<dyn MetaSource>>,
+    initial_fetch_waiter: Waiter,
+}
+
+#[derive(Clone, Debug, Copy)]
+pub struct TypedMetaValue<T> {
+    pub value: T,
+    /// None means unknown revision
+    pub revision: Option<u64>,
+    pub priority: u64,
+}
+
+impl MetaService {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Get the value for the meta field.
+    ///
+    /// This may wait for significant time on first run.
+    pub async fn get_field<V: DeserializeOwned + 'static>(
+        &self,
+        db: &Database,
+        field: &str,
+    ) -> Option<TypedMetaValue<V>> {
+        if let Some(value) = self.get_field_inner(db, field).await {
+            // might be from in old cache.
+            // TODO: maybe old cache should have a ttl?
+            Some(value)
+        } else {
+            // wait for initial value
+            self.initial_fetch_waiter.wait().await;
+            self.get_field_inner(db, field).await
+        }
+    }
+
+    async fn get_field_inner<V: DeserializeOwned + 'static>(
+        &self,
+        db: &Database,
+        field: &str,
+    ) -> Option<TypedMetaValue<V>> {
+        db.begin_transaction_nc()
+            .await
+            // sorted by priority
+            .find_by_prefix_sorted_descending(&MetaFieldName(field.to_string()))
+            .await
+            .collect::<Vec<_>>()
+            .await
+            .into_iter()
+            .find_map(|(key, value)| {
+                Some(TypedMetaValue {
+                    revision: value.revision,
+                    value: parse_meta_value_static::<V>(&value.value).ok()?,
+                    priority: key.priority,
+                })
+            })
+    }
+
+    pub fn add_source(&mut self, source: impl MetaSource + 'static) {
```

Comment:
IMO, mutability here is asking for troubles. The whole thing should start with all sources known and then not require/support modifications. Possibly a builder can be used to assemble the sources, then convert to a ready service.

Reply:
given `update_continuously` is `(&self) -> !` it forces the manager to be immutable